Compatibility with Python 3.12: AssertionError: RuntimeError not raised by bdecode

Hello, 

I'd like to ask for help. In Fedora, we are using the upcoming Python 3.12 as the main Python in Fedora Rawhide. We are seeing the following test failure:

~~~~
======================================================================
FAIL: test_list_deepnested (fastbencode.tests.test_bencode.TestBencodeDecode.test_list_deepnested)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/fastbencode-0.0.9/fastbencode/tests/test_bencode.py", line 297, in test_list_deepnested
    self._run_check_error(RuntimeError, (b"l" * 100) + (b"e" * 100))
  File "/builddir/build/BUILD/fastbencode-0.0.9/fastbencode/tests/test_bencode.py", line 238, in _run_check_error
    self.assertRaises(exc, self.module.bdecode, bad)
AssertionError: RuntimeError not raised by bdecode
~~~~

--- a/fastbencode/tests/test_bencode.py
+++ b/fastbencode/tests/test_bencode.py
@@ -287,10 +287,16 @@ class TestBencodeDecode(TestCase):
 
     def test_list_deepnested(self):
         import platform
-        if platform.python_implementation() == 'PyPy':
-            self.skipTest('recursion not an issue on pypy')
-        with RecursionLimit():
-            self._run_check_error(RuntimeError, (b"l" * 100) + (b"e" * 100))
+        if (platform.python_implementation() == 'PyPy'
+                or sys.version_info[:2] >= (3, 12)):
+            expected = []
+            for i in range(99):
+                expected = [expected]
+            self._check(expected, (b"l" * 100) + (b"e" * 100))
+        else:
+            with RecursionLimit():
+                self._run_check_error(
+                    RuntimeError, (b"l" * 100) + (b"e" * 100))
 
     def test_malformed_list(self):
         self._run_check_error(ValueError, b'l')

